src/collision.cpp

```diff
@@ -301,12 +301,12 @@ static void add_object_boxes(Environment *env,
 
 		// add collision with local player
 		LocalPlayer *lplayer = c_env->getLocalPlayer();
-		if (lplayer->getParent() == nullptr) {
+		auto *obj = (ActiveObject*) lplayer->getCAO();
+		if (self != obj && !lplayer->getParent()) {
```

Comment:
Confirmed that LocalPlayer collides with itself, using this patch that makes walking on air possible:

```diff
diff --git a/src/collision.cpp b/src/collision.cpp
index d804f7f0d..01738cadb 100644
--- a/src/collision.cpp
+++ b/src/collision.cpp
@@ -306,6 +306,8 @@ static void add_object_boxes(Environment *env,
                        v3f lplayer_pos = lplayer->getPosition();
                        lplayer_collisionbox.MinEdge += lplayer_pos;
                        lplayer_collisionbox.MaxEdge += lplayer_pos;
+                       lplayer_collisionbox.MaxEdge.Y = lplayer_collisionbox.MinEdge.Y;
+                       lplayer_collisionbox.MinEdge.Y -= 0.5f;
                        auto *obj = (ActiveObject*) lplayer->getCAO();
                        cinfo.emplace_back(obj, 0, lplayer_collisionbox);
                }
```

Reply:
I propose these changes: https://github.com/minetest/minetest/commit/1a442e6da67c7ef62ef1d615db9af295ad856541. This fixes the self-collision issue in both cases.

Reply:
Currently it seems to me that the parameter `self` specifies the ActiveObject that must be ignored to avoid colliding with itself. This PR extends this idea for LocalPlayer's CAO.

Counter-proposal:
```diff
diff --git a/src/client/localplayer.cpp b/src/client/localplayer.cpp
index 51e91ab94..75f755578 100644
--- a/src/client/localplayer.cpp
+++ b/src/client/localplayer.cpp
@@ -360,7 +360,7 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
 
        collisionMoveResult result = collisionMoveSimple(env, m_client,
                pos_max_d, m_collisionbox, player_stepheight, dtime,
-               &position, &m_speed, accel_f);
+               &position, &m_speed, accel_f, m_cao);
 
        bool could_sneak = control.sneak && !free_move && !in_liquid &&
                !is_climbing && physics_override.sneak;
@@ -990,7 +990,7 @@ void LocalPlayer::old_move(f32 dtime, Environment *env, f32 pos_max_d,
 
        collisionMoveResult result = collisionMoveSimple(env, m_client,
                pos_max_d, m_collisionbox, player_stepheight, dtime,
-               &position, &m_speed, accel_f);
+               &position, &m_speed, accel_f, m_cao);
 
        // Position was slightly changed; update standing node pos
        if (touching_ground)
@@ -1254,7 +1254,7 @@ void LocalPlayer::handleAutojump(f32 dtime, Environment *env,
 
        // try at peak of jump, zero step height
        collisionMoveResult jump_result = collisionMoveSimple(env, m_client, pos_max_d,
-               m_collisionbox, 0.0f, dtime, &jump_pos, &jump_speed, v3f(0.0f));
+               m_collisionbox, 0.0f, dtime, &jump_pos, &jump_speed, v3f(0.0f), m_cao);
 
        // see if we can get a little bit farther horizontally if we had
        // jumped
```

Reply:
👍 for your proposal

In two other places, the check is `self != obj && self != obj->getParent()`. Would the second part also make sense here?

Reply:
You're right. `self` must not collide with itself and not with its children.